File: solutions/python/tournament/1/tournament.py
import string


class SoccerTeam:

    def __init__(self, name):
        self.name = name
        self.mp = 0
        self.w = 0
        self.d = 0
        self.l = 0
        self.p = 0

    def win(self):
        self.mp += 1
        self.w += 1
        self.p += 3

    def draw(self):
        self.mp += 1
        self.d += 1
        self.p += 1

    def loss(self):
        self.mp += 1
        self.l += 1

    def __str__(self):
        str_mp = str(self.mp).rjust(2)
        str_w = str(self.w).rjust(2)
        str_d = str(self.d).rjust(2)
        str_l = str(self.l).rjust(2)
        str_p = str(self.p).rjust(2)
        return f"{self.name.ljust(30)} | {str_mp} | {str_w} | {str_d} | {str_l} | {str_p}"


def tally(rows):
    teams = {}
    for item in rows:
        p1, p2, r = item.split(';')
        if not p1 in teams:
            teams[p1] = SoccerTeam(p1)
        if not p2 in teams:
            teams[p2] = SoccerTeam(p2)
        if r == 'win':
            teams[p1].win()
            teams[p2].loss()
        if r == 'draw':
            teams[p1].draw()
            teams[p2].draw()
        if r == 'loss':
            teams[p1].loss()
            teams[p2].win()
    team_list = list(teams.values())
    # team_list.sort(reverse=True, key=lambda t1, t2: (t1.name < t2.name) if t1.p == t2.p else (t1.p == t2.p))
    team_list.sort(reverse=False, key=lambda t: (- t.p, t.name))
    return ["Team                           | MP |  W |  D |  L |  P"] + [str(item) for item in team_list ]
        

File: solutions/java/tournament/1/src/main/java/Tournament.java
import java.util.HashMap;

class Tournament {

    private HashMap<String, Team> map = new HashMap<>();

    String printTable() {
        StringBuilder sBuilder = new StringBuilder();
        sBuilder.append("Team                           | MP |  W |  D |  L |  P\n");
        for (Team s: this.map.values().stream().sorted().toList()) {
            sBuilder.append(s.toString() + "\n");
        }
        return sBuilder.toString();
    }

    void applyResults(String resultString) {
        String[] rows = resultString.split("\n");
        for (String r: rows) {
            String [] cols = r.split(";");
            String team0 = cols[0], team1 = cols[1], result = cols[2];
            if (!this.map.containsKey(team0)) { this.map.put(team0, new Team(team0)); }
            if (!this.map.containsKey(team1)) { this.map.put(team1, new Team(team1)); }
            if (result.equals("win")) {
                this.map.get(team0).win();
                this.map.get(team1).lost();
            }
            else if (result.equals("loss")) {
                this.map.get(team0).lost();
                this.map.get(team1).win();
            }
            else if (result.equals("draw")) {
                this.map.get(team0).draw();
                this.map.get(team1).draw();
            }
        }
    }


    private class Team implements Comparable<Team>{
        
        final String name;
        public int points;
        public int winGames;
        public int lostGames;
        public int drawGames;
        public int playedGames;

        public Team(String name) {
            this.name = name;
            this.points = 0;
            this.winGames = 0;
            this.lostGames = 0;
            this.drawGames = 0;
            this.playedGames = 0;
        }

        public void win() {
            this.points += 3;
            this.playedGames++;
            this.winGames++;
        }

        public void draw() {
            this.points += 1;
            this.drawGames++;
            this.playedGames++;
        }

        public void lost() {
            this.lostGames++;
            this.playedGames++;
        }

        @Override
        public String toString() {
            return String.format(
                "%-30s | %2d | %2d | %2d | %2d | %2d",
                name,           // %-30s → alineado a la izquierda, 30 caracteres
                playedGames,    // %2d → 2 dígitos alineados a la derecha
                winGames,
                drawGames,
                lostGames,
                points
            );
        }

        @Override
        public int compareTo(Team t) {
            int diff = t.points - this.points;
            if (diff != 0) { return diff; }
            return this.name.compareTo(t.name);
        }

    }
}
